pyproject.toml

```diff
@@ -7,12 +7,123 @@ requires = [
 ]
 build-backend = "setuptools.build_meta"
 
+[project]
+name = "rasterio"
+dynamic = ["version"]
+authors = [
+    {name = "Sean Gillies", email = "sean@mapbox.com"},
+]
+maintainers = [
+    {name = "Rasterio contributors"},
+]
+description = "Fast and direct raster I/O for use with NumPy and SciPy"
+readme = "README.rst"
+keywords = ["gis", "raster", "gdal"]
+license = "BSD-3-Clause"
+license-files = ["LICENSE.txt", "AUTHORS.txt"]
+classifiers = [
+    "Development Status :: 5 - Production/Stable",
+    "Intended Audience :: Developers",
+    "Intended Audience :: Information Technology",
+    "Intended Audience :: Science/Research",
+    "Programming Language :: C",
+    "Programming Language :: Cython",
+    "Programming Language :: Python :: 3.9",
+    "Programming Language :: Python :: 3.10",
+    "Programming Language :: Python :: 3.11",
+    "Programming Language :: Python :: 3.12",
+    "Programming Language :: Python :: 3.13",
+    "Programming Language :: Python :: 3.14",
+    "Programming Language :: Python :: 3",
+    "Topic :: Multimedia :: Graphics :: Graphics Conversion",
+    "Topic :: Scientific/Engineering :: GIS",
+]
+requires-python = ">=3.9"
+dependencies = [
+    "affine",
+    "attrs",
+    "certifi",
+    # Avoid pallets/click#2939.
+    "click>=4.0,!=8.2.*",
+    "cligj>=0.5",
+    "importlib-metadata ; python_version < '3.10'",
+    "numpy>=1.24",
+    "click-plugins",
+    "pyparsing",
+]
+
+[project.optional-dependencies]
+all = ["rasterio[docs,ipython,plot,s3,test]"]
+docs = [
+    "ghp-import",
+    "numpydoc",
+    "sphinx",
+    "sphinx-click",
+    "sphinx-rtd-theme",
+]
+ipython = ["ipython>=2.0"]
+plot = ["matplotlib"]
+s3 = ["boto3>=1.2.4"]
+test = [
+    "aiohttp",
+    "boto3>=1.2.4",
+    "fsspec",
+    "hypothesis",
+    "packaging",
+    "pytest-cov>=2.2.0",
+    "pytest>=2.8.2",
+    "requests",
+    "shapely",
+]
```

Comment:
I think a lot of these may fit better into [`[dependency-groups]`](https://packaging.python.org/en/latest/specifications/dependency-groups/)
rather than `[optional-dependencies]`.

Groups fits better dependencies used in development stage (thinks like pytest, ipython etc.) for which you are expected users to clone the repo and install the env from source.

In contrast optional-dependencies contains deps that are used in the source code of a package and needs to be shipped with the .whl